src/librustdoc/clean/types.rs

```diff
@@ -743,67 +741,122 @@ impl Item {
+    fn repr<'tcx>(&self, tcx: TyCtxt<'tcx>, cache: &Cache) -> Option<String> {
+        let def_id = self.def_id()?;
+        let (ItemType::Struct | ItemType::Enum | ItemType::Union) = self.type_() else {
+            return None;
+        };
+
+        let adt = tcx.adt_def(def_id);
+        let repr = adt.repr();
+
+        let is_visible = |def_id| cache.document_hidden || !tcx.is_doc_hidden(def_id);
+        let is_field_public =
+            |field: &'tcx ty::FieldDef| field.vis.is_public() && is_visible(field.did);
+
+        if repr.transparent() {
+            // `repr(transparent)` is public iff the non-1-ZST field is public or
+            // at least one field is public in case all fields are 1-ZST fields.
+            let is_public = cache.document_private
+                || adt.variants().iter().all(|variant| {
+                    if !is_visible(variant.def_id) {
+                        return false;
                     }
-                    IntegerType::Fixed(size, is_signed) => {
-                        format!("{}{}", if is_signed { 'i' } else { 'u' }, size.size().bytes() * 8)
+
+                    let field = variant.fields.iter().find(|field| {
+                        let args = ty::GenericArgs::identity_for_item(tcx, field.did);
+                        let ty = field.ty(tcx, args);
+                        tcx.layout_of(tcx.param_env(field.did).and(ty))
+                            .is_ok_and(|layout| !layout.is_1zst())
+                    });
+
+                    if let Some(field) = field {
+                        return is_field_public(field);
                     }
-                };
-                out.push(&int_s);
-            }
-            if !out.is_empty() {
-                attrs.push(format!("#[repr({})]", out.join(", ")));
-            }
+
+                    adt.variants().iter().all(|variant| {
+                        variant.fields.is_empty() || variant.fields.iter().any(is_field_public)
+                    })
+                });
+
+            // Since `repr(transparent)` can't have any other reprs or
+            // repr modifiers beside it, we can safely return early here.
+            return is_public.then(|| "#[repr(transparent)]".into());
         }
-        attrs
+
+        // Fast path which avoids looking through the variants and fields in
+        // the common case of no `#[repr]` or in the case of `#[repr(Rust)]`.
+        if !repr.c()
+            && !repr.simd()
+            && repr.int.is_none()
+            && repr.pack.is_none()
+            && repr.align.is_none()
+        {
+            return None;
+        }
+
+        let is_public = cache.document_private
+            || if adt.is_enum() {
+                // FIXME(fmease): Should we take the visibility of fields of variants into account?
+                // FIXME(fmease): `any` or `all`?
+                adt.variants().is_empty()
```

Comment:
I think I'd go for `any` in both cases.

Reply:
@RalfJung, does

```js
public($REPR)
    <=>
ANY_OR_EMPTY V: Variants($ADT).
    ANY_OR_EMPTY F: Fields(V).
        public(F)
```

sound good to you as well? Or should it be `ANY_OR_EMPTY ALL` (current impl) or `ALL ALL` or `ALL ANY_OR_EMPTY`.

Reply:
Usually for structs, if there is at least one private field then we say you can't rely on the struct staying how it is. For instance if your `repr(transparent)` relies on another type being a ZST and that type has at least one private field, we warn about that (and we eventually want to make that an error).

So I'd say the same should go for the repr. If any field is private, then the repr is private.

Reply:
There's no such things as private enum variants (unfortunately).

Reply:
Oops, I meant `doc(hidden)` variants ^^'

Reply:
I'm not sure which rules people usually use for when an enum is not-semver-stable. `#[non_exhaustive]` is one thing. (A `#[non_exhaustive]`  struct should probably also have its `repr` hidden.) All fields of all variants are always public so that's probably the only thing one can do?

Reply:
> Oops, I meant doc(hidden) variants ^^'

Ah. that's an attribute I don't have intuition for, sorry. In rustc it doesn't affect anything.

Can one add this to a field? So on a struct with a doc(hidden) field, should the repr be shown? No idea.

Reply:
> on a struct with a doc(hidden) field, should the repr be shown?

Here is an example of what goes wrong if you show repr on a struct with doc(hidden) field.

```rust
#[repr(C)]
pub struct Struct {
    pub first: u8,
    #[doc(hidden)]
    pub second: u16,
    pub third: u8,
}
```

<img width="506" alt="Screenshot 2023-10-18 at 3 41 44 PM" src="https://github.com/rust-lang/rust/assets/1940490/34b439ab-5d62-4cad-9703-e0a8e8f01fa2">

Rustdoc purports a repr(C) struct in which the first byte is `first`, the second byte is `third`, and some other fields follow. Given the Rust-like syntax in which rustdoc shows #\[repr(C)\], this feels misleading. For a struct that is _actually_ this:

```rust
#[repr(C)]
pub struct Struct {
    pub first: u8,
    pub third: u8,
    // ... other fields ...
}
```

one would expect they can cast `&Struct` to `&[u8; 2]` and read `first` and `third` from it. If they do that in this case though, they get UB from looking at a padding byte.

I think this would be a useful bar to keep in mind as a minimum desirable property; rustdoc should not show a repr in such a way that misleads reader about reality. That does not necessarily need to mean hiding such reprs, though that might be the most expedient path forward. Alternatively rustdoc could be more discerning about placing the /\*private field\*/ comment in between the correct fields when there is a repr.

Reply:
Hmm, do we already track this in a GitHub issue? With the (unstable) introduction of `std::mem::offset_of`, it feels like we should up the priority of it. If I remember correctly, it'd need quite a bit of rewiring inside rustdoc to render `/* private field */` in the correct order for `repr(C)` structs since those fields are *stripped* early.

Reply:
RalfJung, re taking `doc(hidden)` on variants into account when computing the visibility of a `repr`, I've included that to hide the `repr(u8)` on `core::ffi::c_void` which has two `doc(hidden)` enum variants.

Reply:
Sounds to me then like we'd want to hide the `repr` as soon as there is any hidden field (just like we hide it as soon as there is any private field).

Reply:
Seems like there is more to discuss. I'll add it to the next [rustdoc team meeting agenda](https://hackmd.io/ITqOnVbCQaWJzHm5unTjEQ?both).